ambry-clustermap/src/main/java/com.github.ambry.clustermap/Replica.java

```diff
@@ -29,27 +31,48 @@
  * its {@link Disk}. Note that this induces a constraint that a Partition can never have more than one Replica on a
  * given Disk. This ensures that a Partition does not have Replicas that share fates.
  */
-class Replica implements ReplicaId {
+class Replica implements ReplicaId, Resource {
   private final Partition partition;
   private Disk disk;
   private volatile boolean isStopped = false;
+  private final ResourceStatePolicy resourceStatePolicy;
 
   private Logger logger = LoggerFactory.getLogger(getClass());
 
-  Replica(Partition partition, Disk disk) {
+  Replica(Partition partition, Disk disk, ClusterMapConfig clusterMapConfig) {
     if (logger.isTraceEnabled()) {
       logger.trace("Replica " + partition + ", " + disk);
     }
     this.partition = partition;
     this.disk = disk;
-
+    try {
+      ResourceStatePolicyFactory resourceStatePolicyFactory =
+          Utils.getObj(clusterMapConfig.clusterMapResourceStatePolicyFactory, this, HardwareState.AVAILABLE,
+              clusterMapConfig);
+      resourceStatePolicy = resourceStatePolicyFactory.getResourceStatePolicy();
+    } catch (Exception e) {
+      logger.error("Error creating resource state policy when instantiating a replica " + e);
+      throw new IllegalStateException("Error creating resource state policy when instantiating a replica " + partition,
+          e);
+    }
     validate();
   }
 
   Replica(HardwareLayout hardwareLayout, Partition partition, JSONObject jsonObject) throws JSONException {
     this.partition = partition;
```

Comment:
if possible, call the other constructor in this one
`this(partition, hardwareLayout.findDisk(jsonObject.getString("hostname"), jsonObject.getInt("port"), jsonObject.getString("mountPath")),  hardwareLayout.getClusterMapConfig())`